### PR TITLE
fix: dsl encoding detection enhanced.

### LINKED ISSUE #1278: [Bug] Failure to index some .dsl dictionaries (UTF16 without BOM)
**Describe the bug**
Some .dsl files can be added to the list of dictionaries, but the software fails to index the headwords. So it loads a "ghost" dictionary with no Title and no headwords. 

**To Reproduce**
1. Download a sample of 5 German Dictionaries in .dsl format: 
https://nch.pl/s/6L7r9caNby4n8LJ
3. Try to index those dictionaries
4. GoldenDict shows "0 headwords" and no Title for those dictionaries.

**Affected Dictionaries**

At least 10 dictionaries (English and German). Many of them are made by myself but not all!

**Download:** https://nch.pl/s/6L7r9caNby4n8LJ

**Expected behavior**
All those .dsl dictionaries should work normally. GoldenDict-ng was able to recognize them few months ago. 

**Screenshots**

image

image

**OS and software versions**
_Goldendict-ng 23.10.01.2799b41c at 2023-11-06T18:24:56
Qt 6.6.0 GCC 13.2.1 20230801 endeavouros linux 6.6.1-arch1-1 x86_64-little_endian-lp64
Flags:USE_XAPIAN MAKE_ZIM_SUPPORT USE_ICONV MAKE_CHINESE_CONVERSION_SUPPORT_

image

**Additional context**
The .dsl files were working correctly a couple of months ago. Then , I changed from Manjaro OS to EndeavourOS (Arch Linux) and installed the git version from AUR. Suddenly, around 10 .dsl dictionaries are unusable. I presume that the problem lies in the commits of the last couple of months of GD.

I am not sure if the issue can be reproduced in other OSes apart from EndeavourOS. I presume that yes. 

**Could you please try to reproduce this bug ?**

Thanks for your great work! :heart:

### PR DESCRIPTION
**Pull request by xiaoyifang**
fix #1278

### CODE REVIEW

File: src/dict/dsl_details.cc [hunk 1]
```diff
@@ -832,7 +832,34 @@ DslScanner::DslScanner( string const & fileName ):
...
   QByteArray ba = QByteArray::fromRawData( (const char *)firstBytes, 50 );
-  codec         = QTextCodec::codecForUtfText( ba, QTextCodec::codecForName( "UTF-8" ) );
+  codec         = QTextCodec::codecForUtfText( ba, nullptr );
+  if ( !codec ) {
+    // the encoding has no bom.
+    // check the first char # (0x23).
+    auto hashTag      = 0x0023;
+    const uchar * buf = reinterpret_cast< const uchar * >( ba.constData() );
+
+    uint uci = qFromUnaligned< uint >( buf );
```
Comment (shenlebantongying) [inline src/dict/dsl_details.cc:842]: `uint` and `ushort` are not in either C or C++'s standards. Maybe replace them with`uint64_t`,`uint32_t` from `<cstdint>`?

**Comment by shenlebantongying** [comment]
Do we really need the conversions? I find the code below is equivalent.
```
auto *utf16_be=new QByteArray("\x00\x23",2);
auto *utf32_le=new QByteArray("\x23\x00\x00\x00",4);

qDebug() << (utf16_be->at(0) == '\x00' && utf16_be->at(1) == '\x23');
qDebug() << (utf32_le->at(0) == '\x23' && utf32_le->at(1) == '\x00' && utf32_le->at(2) == '\x00' && utf32_le->at(3) == '\x00');
```

**Comment by xiaoyifang** [comment]
> Do we really need the conversions? I find the code below is equivalent.
> 
> ```
> auto *utf16_be=new QByteArray("\x00\x23",2);
> auto *utf32_le=new QByteArray("\x23\x00\x00\x00",4);
> 
> qDebug() << (utf16_be->at(0) == '\x00' && utf16_be->at(1) == '\x23');
> qDebug() << (utf32_le->at(0) == '\x23' && utf32_le->at(1) == '\x00' && utf32_le->at(2) == '\x00' && utf32_le->at(3) == '\x00');
> ```

I just copy the implementation from codecForUtfText  .

**Comment by shenlebantongying** [comment]
> I just copy the implementation from QTextCodec::codecForUtfText .

They used the comparison that I suggested in the past.

https://github.com/qt/qtbase/commit/99baa0d4401516cff978cab8a7102fc0a25731a9

Then they changed it 5 years ago to the conversion method for some reason compiler can be a better job.

(Instead of accessing bytes one by one, just give cpu a number, then compare it with another number to return as fast as possible? I don't really know.)

**Comment by xiaoyifang** [comment]
from the perspective of hiding details ,  I think the later is better .

**Pull request merged**